### PR TITLE
[14.0] base_rest: improve service context and security

### PR DESCRIPTION
**Pull request by lmignon**
Improvements:

* This PR introduces a new component `BaseRestServiceContextProvider`to be used to extend the component context that will be provided to the service layer when a method is called. Prior to this change, the only way to extend the context was to extend the default method `_get_component_context` provided by the RestController class. Since only one RestController should be registered into the database for a root_path and collection name, this mechanism give you the opportunity to extend the context provided by a base addon into specialized addons
*  If 2 `RestController`s are registered for the same root_path and  collection name, an error is now printed into the log at server startup
* It's now possible to output into the openapi doc, the authentication mechanism required by a method.
* The management of defaults route attributes defined on the RestController has been improved (applied only if not provided by a `@http.route` nor `|restapi.method`decorators)  
* Add a new kind of restapi method parameter `BinaryData`to be used to decalre method requesting/returning files

### CODE REVIEW

File: base_rest/controllers/main.py [hunk 1]
```diff
@@ -44,48 +44,81 @@ def __init__(cls, name, bases, attrs):  # noqa: B902
...
+    controller generated from the service components.
+
+    You must inherit of this controller into your code to register the root path
+    used to serve all the services defined for the given the collection name.
+    your REST. This registration requires 2 parameters:
 
     _root_path:
     _collection_name:
 
+    Only one controller by _collection_name, _root_patch should exists into an
```
Comment (simahawk) [inline base_rest/controllers/main.py]:
```suggestion
    Only one controller by _collection_name, _root_path should exists into an
```

File: base_rest/controllers/main.py [hunk 2]
```diff
@@ -44,48 +44,81 @@ def __init__(cls, name, bases, attrs):  # noqa: B902
...
+    odoo database. If more than one controller exists, a warning is issued into
+    the log at startup and the concrete controller used as base class
+    for the services registered into the collection name and served at the
+    root path is not predictable.
```
Comment (simahawk) [inline base_rest/controllers/main.py:79]: this is not clear, can you rephrase?

File: base_rest/controllers/main.py [hunk 3]
```diff
@@ -44,48 +44,81 @@ def __init__(cls, name, bases, attrs):  # noqa: B902
...
+
+    The following properties can be specified to define comman properties to
+    applies to the generetated REST routes.
```
Comment (simahawk) [inline base_rest/controllers/main.py]:
```suggestion
    The following properties can be specified to define common properties to
    apply to generated REST routes.
```

File: base_rest/models/rest_service_registration.py [hunk 4]
```diff
@@ -107,15 +114,30 @@ def _update_auth_method_controller(self, controller_class):
...
+    def _apply_default_cors_if_not_set(self, controller_class, routing):
+        default_attr_name = "_default_cors"
+        if hasattr(controller_class, default_attr_name) and "cors" not in routing:
+            cors = getattr(controller_class, default_attr_name)
+            routing["cors"] = cors
+            if cors:
+                # add http method 'OPTIONS' required by cors if the route is
+                # restricted to specific method
+                if routing.get("methods"):
+                    routing["methods"].append("OPTIONS")
```
Comment (simahawk) [inline base_rest/models/rest_service_registration.py]:
```suggestion
            if cors and "OPTIONS" not in routing.get("methods", []):
                # add http method 'OPTIONS' required by cors if the route is
                # restricted to specific method
                routing["methods"].append("OPTIONS")
```

**Review by simahawk — APPROVED** [review]
just code review, LG

File: base_rest/models/rest_service_registration.py (hunk 4 above)
Comment (lmignon) [reply base_rest/models/rest_service_registration.py]: `methods`is not required. If not specified, all the methods are allowed for the given route...
```if cors and "OPTIONS" not in routing.get("methods", ["OPTIONS"]):```

File: base_rest/controllers/main.py (hunk 2 above)
Comment (lmignon) [reply base_rest/controllers/main.py:79]: Is-it clearer now? (btw thank you for the review)

**Comment by simahawk** [comment]
can you squash?

**Comment by lmignon** [comment]
@simahawk squash done..

**Comment by simahawk** [comment]
good to merge?

**Comment by lmignon** [comment]
IMO yes @sbidoul, is it ok for you?

File: base_rest/models/rest_service_registration.py [hunk 5]
```diff
@@ -107,15 +109,29 @@ def _update_auth_method_controller(self, controller_class):
...
+        default_attr_name = "_default_" + attr_name
+        if hasattr(controller_class, default_attr_name) and attr_name not in routing:
+            routing[attr_name] = getattr(controller_class, default_attr_name)
+
+    def _apply_default_cors_if_not_set(self, controller_class, routing):
+        default_attr_name = "_default_cors"
+        if hasattr(controller_class, default_attr_name) and "cors" not in routing:
+            cors = getattr(controller_class, default_attr_name)
+            routing["cors"] = cors
+            if cors and "OPTIONS" not in routing.get("methods", ["OPTIONS"]):
```
Comment (sbidoul) [inline base_rest/models/rest_service_registration.py:136]: why `routing.get("methods", ["OPTIONS"])` and not `routing.get("methods", [])` ?

Comment (lmignon) [reply base_rest/models/rest_service_registration.py:136]: because if `methods` is not specified, all methods are allowed... https://github.com/odoo/odoo/blob/13.0/odoo/http.py#L441

File: base_rest/restapi.py [hunk 6]
```diff
@@ -50,18 +50,16 @@ def decorator(f):
                 paths = [paths]
             if not isinstance(http_methods, list):
                 http_methods = [http_methods]
-            if cors and "OPTIONS" not in http_methods:
+            if kw.get("cors", False) and "OPTIONS" not in http_methods:
```
Comment (sbidoul) [inline base_rest/restapi.py]: `kw.get("cors")`

File: base_rest/restapi.py [hunk 7]
```diff
@@ -5,15 +5,13 @@
...
+from odoo import _, http
 from odoo.exceptions import UserError
 
 from .tools import cerberus_to_json
 
 
-def method(
-    routes, input_param=None, output_param=None, auth=None, cors=None, csrf=False
-):
+def method(routes, input_param=None, output_param=None, **kw):
```
Comment (sbidoul) [inline base_rest/restapi.py:14]: explicit auth, cors, csrf made them easily discoverable.

Comment (lmignon) [reply base_rest/restapi.py:14]: It's on purpose... It's the only way I found to know if these parameters have been explicitly provided or not. This information is required to know if the default parameters provided by the controller must be applied or not on a route.

**Review by sbidoul — APPROVED** [review]

File: base_rest/restapi.py (hunk 7 above)
Comment (sbidoul) [reply base_rest/restapi.py:14]: ah, ok, thanks
